File: javascript/bob/bob.js
var Bob = function() {};

Bob.prototype.hey = function(input) {
    if (isShouting(input)) {
        return 'Whoa, chill out!';
    } else if (isQuestion(input)) {
        return 'Sure.';
    } else if (isSilent(input)) {
        return 'Fine. Be that way!';
    } else {
        return 'Whatever.';
    }
};

function isShouting(input) {
    return input.toUpperCase() === input && input.toLowerCase() !== input.toUpperCase();
}

function isQuestion(input) {
    return input.charAt(input.length-1) === '?';
}

function isSilent(input) {
    return input.trim().length === 0;
}

module.exports = Bob;

File: scala/bob/src/main/scala/Bob.scala
class Bob {
  def hey(message: String): String = message match {
    case Shouting() => "Whoa, chill out!"
    case Silence() => "Fine. Be that way!"
    case Question() => "Sure."
    case _ => "Whatever."
  }

  case object Silence {
    def unapply(message: String) = message.trim.isEmpty
  }

  case object Question {
    def unapply(message: String) = message.last == '?'
  }

  case object Shouting {
    def unapply(message: String) = message.exists(_.isUpper) && !message.exists(_.isLower)
  }

}
